artiq/coredevice/ad53xx.py

```diff
@@ -0,0 +1,328 @@
+
+
+@portable
+def ad53xx_cmd_read_ch(channel, op):
+    """Returns the word that must be written to the DAC to read a given
+    DAC channel register.
+
+    :param channel: DAC channel to read (8 bits)
+    :param op: The channel register to read, one of
+      :const:`AD53XX_CMD_DATA`, :const:`AD53XX_CMD_OFFSET` or
+      :const:`AD53XX_CMD_GAIN`
+    :return: The 24-bit word to be written to the DAC, aligned as the 24 MSB of
+      a 32-bit integer, ready to be transferred directly by the SPI core.
+    """
+    return (AD53XX_CMD_SPECIAL | AD53XX_SPECIAL_READ | op |
+           (((channel & 0x3f) + 8) << 7)) << 8
+
+
+@portable
+def voltage_to_mu(voltage, offset_dacs=8192, vref=5.):
+    """Returns the DAC register value required to produce a given output
+    voltage, assuming offset and gain errors have been trimmed out.
+
+    :param voltage: Voltage
+    :param offset_dacs: Register value for the two offset DACs (default
+      :0x1555)
+    :param vref: DAC reference voltage (default: 5.)
+    """
+    return int(round(0x10000*(voltage/(4.*vref)) + offset_dacs*0x4))
+
+@portable
+def offset_to_mu(voltage, offset_dacs=8192, vref=5.):
+    """Returns the offset register value required to produce a given voltage
+    when the DAC register is set to mid-scale.
+
+    An offset of V can be used to trim out a DAC offset error of -V.
+
+    :param voltage: Offset voltage
+    :param offset_dacs: Register value for the two offset DACs (default
+      :0x1555)
+    :param vref: DAC reference voltage (default: 5.)
+    """
+    return int(round(0x10000*(voltage/(4.*vref)) + offset_dacs*0x4))
+
+class AD53xx:
+    """Analog devices AD53[67][0123] family of multi-channel Digital to Analog
+    Converters.
+
+    :param spi_device: SPI bus device name
+    :param ldac_device: LDAC RTIO TTLOut channel name
+    :param clr_device: CLR RTIO TTLOut channel name
+    :param chip_select: Value to drive on SPI chip select lines during
+      transactions (default: 1)
+    :param div_write: SPI clock divider for write operations (default: 4,
+      50MHz max SPI clock with {t_high, t_low} >=8ns)
+    :param div_read: SPI clock divider for read operations (default: 8, not
+      optimized for speed, but cf data sheet t22: 25ns min SCLK edge to SDO
+      valid)
+    :param vref: DAC reference voltage (default: 5.)
+    :param offset_dacs: Initial register value for the two offset DACs, device
+      dependent and must be set correctly for correct voltage to mu conversions
+      (default: 8192)
+    :param core_device: Core device name (default: "core")
+    """
+    kernel_invariants = {"bus", "ldac", "clr", "chip_select", "div_write",
+                         "div_read", "vref", "core"}
+
+    def __init__(self, dmgr, spi_device, ldac_device, clr_device,
+                 chip_select=1, div_write=4, div_read=8, vref=5.,
+                 offset_dacs=8192, core="core"):
+        self.bus = dmgr.get(spi_device)
+        self.ldac = dmgr.get(ldac_device)
```

Comment:
@hartytp Did you eventually deem it unnecessary to support configurations without LDAC?

Reply:
True that works, but it seems a but heavyweight for me unless there is a concrete use case for it - I'm not aware of one. My approach is to start simple and add that kind of feature when a need is expressed.

If you want to add this during merge I don't object.

Reply:
Just need to check that e.g. NIST don't need it... @dhslichter ?

Reply:
Assuming no major objections, would you mind finishing this off from here? This is now all working on the hw and I'd like to prioritize analogy tests on Zotino.

Reply:
Ok.

Reply:
@sbourdeauducq thinking about this more: using this without ldac is probably a rare use case, so let's not clutter up the driver with it if we can avoid it. What about having a general dummy ttl that the device manager can provide for this kind of case?

Reply:
@sbourdeauducq @hartytp we have some DACs that we use without LDAC (it's tied low on the board).  We like having this option because then you free a TTL line for other uses.

Reply:
That option was subsequently added back in.
5ca59467fdfca01e3b0754b1cc6a1d7fbf6ec036

Reply:
Thanks @jordens!